proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,361 @@
+            (Status::Stopped(_since), Event::Protocol(ProtocolEvent::Listening(_addr))) => {
+                self.status = Status::Started(Instant::now());
+
+                vec![]
+            },
+            // Sync with first incoming peer.
+            //
+            // In case the peer is configured to sync on startup we start syncing, otherwise we go
+            // online straight away.
+            (Status::Started(_since), Event::Protocol(ProtocolEvent::Connected(ref peer_id))) => {
+                self.connected_peers.insert(peer_id.clone());
+
+                if self.config.sync_on_startup {
+                    self.status = Status::Syncing(Instant::now(), 1);
+
+                    vec![
+                        Command::SyncPeer(peer_id.clone()),
+                        Command::StartSyncTimeout,
+                    ]
+                } else {
+                    self.status = Status::Online(Instant::now());
+
+                    vec![]
+                }
+            },
+            // Sync until configured maximum of peers is reached.
+            (Status::Syncing(since, syncs), Event::Protocol(ProtocolEvent::Connected(peer_id)))
+                if *syncs < SYNC_MAX_PEERS =>
+            {
+                self.connected_peers.insert(peer_id.clone());
+                if syncs + 1 == SYNC_MAX_PEERS {
+                    self.status = Status::Online(Instant::now());
+                } else {
+                    self.status = Status::Syncing(*since, syncs + 1);
+                }
+
+                vec![Command::SyncPeer(peer_id)]
+            }
+            // TODO(xla): Also issue sync if we come online after a certain period of being
+            // disconnected from any peer.
+            // Issue more syncs if we connect to new peers while syncing.
+            (
+                Status::Syncing(since, syncs),
+                Event::Protocol(ProtocolEvent::Connected(ref peer_id)),
+            ) => {
+                self.status = Status::Syncing(*since, syncs + 1);
+
+                vec![Command::SyncPeer(peer_id.clone())]
+            },
+            // Go online if we exceed the sync period.
+            (Status::Syncing(_since, _syncs), Event::Timeout(TimeoutEvent::SyncPeriod)) => {
+                self.status = Status::Online(Instant::now());
+
+                vec![]
+            },
+            // Go offline if we have no more connected peers left.
+            (_, Event::Protocol(ProtocolEvent::Disconnecting(peer_id)))
+                if self.connected_peers.len() == 1 =>
+            {
+                self.connected_peers.remove(&peer_id);
+                self.status = Status::Offline(Instant::now());
+
+                vec![]
+            }
+            // Remove peer that just disconnected.
+            (_, Event::Protocol(ProtocolEvent::Disconnecting(peer_id))) => {
+                self.connected_peers.remove(&peer_id);
```

Comment:
I think we could merge these two branches:
```rust
self.connect_peers.remove(&peer_id)
if self.connected_peers.is_empty() {
  self.status = Status::Offline(Instant::now())
}
vec![]
```

Reply:
True, tried to be heaps fancy with these guards.

Reply:
And now I'm here to check your work ;)